File: kirun-js/src/engine/json/schema/validator/StringValidator.ts
import { isNullValue } from '../../../util/NullCheck';
import { Schema } from '../Schema';
import { StringFormat } from '../string/StringFormat';
import { SchemaValidationException } from './exception/SchemaValidationException';
import { SchemaValidator } from './SchemaValidator';

export class StringValidator {
    private static readonly TIME: RegExp =
        /^([01]?[0-9]|2[0-3]):[0-5][0-9](:[0-5][0-9])?([+-][01][0-9]:[0-5][0-9])?$/;

    private static readonly DATE: RegExp =
        /^[0-9]{4,4}-([0][0-9]|[1][0-2])-(0[1-9]|[1-2][1-9]|3[01])$/;

    private static readonly DATETIME: RegExp =
        /^[0-9]{4,4}-([0][0-9]|[1][0-2])-(0[1-9]|[1-2][1-9]|3[01])T([01]?[0-9]|2[0-3]):[0-5][0-9](:[0-5][0-9])?([+-][01][0-9]:[0-5][0-9])?$/;

    private static readonly EMAIL: RegExp =
        /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;

    public static validate(parents: Schema[], schema: Schema, element: any): any {
        if (isNullValue(element))
            throw new SchemaValidationException(
                SchemaValidator.path(parents),
                'Expected a string but found ' + element,
            );

        if (typeof element !== 'string')
            throw new SchemaValidationException(
                SchemaValidator.path(parents),
                element.toString() + ' is not String',
            );

        if (schema.getFormat() == StringFormat.TIME) {
            StringValidator.patternMatcher(
                parents,
                schema,
                element,
                StringValidator.TIME,
                'time pattern',
            );
        } else if (schema.getFormat() == StringFormat.DATE) {
            StringValidator.patternMatcher(
                parents,
                schema,
                element,
                StringValidator.DATE,
                'date pattern',
            );
        } else if (schema.getFormat() == StringFormat.DATETIME) {
            StringValidator.patternMatcher(
                parents,
                schema,
                element,
                StringValidator.DATETIME,
                'date time pattern',
            );
        } else if (schema.getFormat() == StringFormat.EMAIL) {
            StringValidator.patternMatcher(
                parents,
                schema,
                element,
                StringValidator.EMAIL,
                'email pattern',
            );
        } else if (schema.getPattern()) {
            StringValidator.patternMatcher(
                parents,
                schema,
                element,
                new RegExp(schema.getPattern()!),
                'pattern ' + schema.getPattern(),
            );
        }

        let length: number = element.length;
        if (schema.getMinLength() && length < schema.getMinLength()!) {
            throw new SchemaValidationException(
                SchemaValidator.path(parents),
                schema.getDetails()?.getValidationMessage('minLength') ??
                'Expected a minimum of ' + schema.getMinLength() + ' characters',
            );
        } else if (schema.getMaxLength() && length > schema.getMaxLength()!) {
            throw new SchemaValidationException(
                SchemaValidator.path(parents),
                schema.getDetails()?.getValidationMessage('maxLength') ??
                'Expected a maximum of ' + schema.getMaxLength() + ' characters',
            );
        }

        return element;
    }

    private static patternMatcher(
        parents: Schema[],
        schema: Schema,
        element: any,
        pattern: RegExp,
        message: string,
    ) {
        let matched: boolean = pattern.test(element);
        if (!matched) {
            throw new SchemaValidationException(
                SchemaValidator.path(parents),
                schema.getDetails()?.getValidationMessage('pattern') ??
                element.toString() + ' is not matched with the ' + message,
            );
        }
    }

    private constructor() {}
}

File: kirun-py/src/kirun_py/json/schema/validator/string_validator.py
from __future__ import annotations
import re
from typing import Any, List, Optional, TYPE_CHECKING

from kirun_py.json.schema.string.string_format import StringFormat
from kirun_py.json.schema.validator.exception.schema_validation_exception import SchemaValidationException
from kirun_py.util.error_message_formatter import ErrorMessageFormatter
from kirun_py.util.null_check import is_null_value

if TYPE_CHECKING:
    from kirun_py.json.schema.schema import Schema

# Pre-compiled format patterns
_TIME_PATTERN = re.compile(
    r'^([01]?[0-9]|2[0-3]):[0-5][0-9](:[0-5][0-9])?(\.\d{1,3})?(Z|[+-][01][0-9]:[0-5][0-9])?$'
)

_DATE_PATTERN = re.compile(
    r'^[0-9]{4}-([0][0-9]|[1][0-2])-(0[1-9]|[1-2][0-9]|3[01])$'
)

_DATETIME_PATTERN = re.compile(
    r'^[0-9]{4}-([0][0-9]|[1][0-2])-(0[1-9]|[1-2][0-9]|3[01])T([01]?[0-9]|2[0-3]):[0-5][0-9](:[0-5][0-9])?(\.\d{1,3})?(Z|[+-][01][0-9]:[0-5][0-9])?$'
)

_EMAIL_PATTERN = re.compile(
    r"^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$"
)


class StringValidator:

    @staticmethod
    def validate(parents: List[Schema], schema: Schema, element: Any) -> Any:
        from kirun_py.json.schema.validator.schema_validator import SchemaValidator

        if is_null_value(element):
            raise SchemaValidationException(
                SchemaValidator.path(parents),
                f'Expected a string but found {ErrorMessageFormatter.format_value(element)}',
            )

        if not isinstance(element, str):
            raise SchemaValidationException(
                SchemaValidator.path(parents),
                f'Expected a string but found {ErrorMessageFormatter.format_value(element)}',
            )

        fmt = schema.get_format()

        if fmt == StringFormat.TIME:
            StringValidator._pattern_matcher(
                parents, schema, element, _TIME_PATTERN, 'time pattern',
            )
        elif fmt == StringFormat.DATE:
            StringValidator._pattern_matcher(
                parents, schema, element, _DATE_PATTERN, 'date pattern',
            )
        elif fmt == StringFormat.DATETIME:
            StringValidator._pattern_matcher(
                parents, schema, element, _DATETIME_PATTERN, 'date time pattern',
            )
        elif fmt == StringFormat.EMAIL:
            StringValidator._pattern_matcher(
                parents, schema, element, _EMAIL_PATTERN, 'email pattern',
            )
        elif fmt == StringFormat.REGEX:
            # For REGEX format, validate that the string is a valid regex
            try:
                re.compile(element)
            except re.error:
                raise SchemaValidationException(
                    SchemaValidator.path(parents),
                    element + ' is not a valid regular expression',
                )
        elif schema.get_pattern():
            try:
                compiled = re.compile(schema.get_pattern())
            except re.error:
                raise SchemaValidationException(
                    SchemaValidator.path(parents),
                    'Invalid regex pattern: ' + schema.get_pattern(),
                )
            StringValidator._pattern_matcher(
                parents, schema, element, compiled, 'pattern ' + schema.get_pattern(),
            )

        length = len(element)

        if schema.get_min_length() is not None and length < schema.get_min_length():
            details = schema.get_details()
            msg = (
                details.get_validation_message('minLength')
                if details else None
            )
            raise SchemaValidationException(
                SchemaValidator.path(parents),
                msg or ('Expected a minimum of ' + str(schema.get_min_length()) + ' characters'),
            )

        if schema.get_max_length() is not None and length > schema.get_max_length():
            details = schema.get_details()
            msg = (
                details.get_validation_message('maxLength')
                if details else None
            )
            raise SchemaValidationException(
                SchemaValidator.path(parents),
                msg or ('Expected a maximum of ' + str(schema.get_max_length()) + ' characters'),
            )

        return element

    @staticmethod
    def _pattern_matcher(
        parents: List[Schema],
        schema: Schema,
        element: Any,
        pattern: re.Pattern,
        message: str,
    ) -> None:
        from kirun_py.json.schema.validator.schema_validator import SchemaValidator

        matched = pattern.search(element) is not None
        if not matched:
            details = schema.get_details()
            msg = (
                details.get_validation_message('pattern')
                if details else None
            )
            raise SchemaValidationException(
                SchemaValidator.path(parents),
                msg or (str(element) + ' is not matched with the ' + message),
            )
